napari/components/_tests/test_layer_slicer.py

```diff
@@ -271,6 +276,97 @@ def test_slice_layers_async_task_to_layers_lock(layer_slicer):
     assert task not in layer_slicer._layers_to_task
 
 
+def test_slice_layers_exception_main_thread(layer_slicer):
+    """Exception is raised on the main thread from an error on the main
+    thread immediately when the task is created."""
+
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        def _make_slice_request(self, dims) -> FakeSliceRequest:
+            raise RuntimeError('_make_slice_request')
+
+    layer = FakeAsyncLayerError()
+    with pytest.raises(RuntimeError, match='_make_slice_request'):
+        layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+
+def test_slice_layers_exception_subthread_on_result(layer_slicer):
+    """Exception is raised on the main thread from an error on a subthread
+    only after result is called, not upon submission of the task."""
+
+    @dataclass(frozen=True)
+    class FakeSliceRequestError(FakeSliceRequest):
+        def __call__(self) -> FakeSliceResponse:
+            assert current_thread() != main_thread()
+            raise RuntimeError('FakeSliceRequestError')
+
+    class FakeAsyncLayerError(FakeAsyncLayer):
+        def _make_slice_request(self, dims) -> FakeSliceRequestError:
+            return FakeSliceRequestError(id=self.slice_count, lock=self.lock)
+
+    layer = FakeAsyncLayerError()
+    future = layer_slicer.slice_layers_async(layers=[layer], dims=Dims())
+
+    done, _ = wait([future], timeout=5)
+    if done:
+        with pytest.raises(RuntimeError, match='FakeSliceRequestError'):
+            future.result()
+    else:
+        raise TimeoutError('Test future did not complete within timeout.')
+
+
+def test_wait_until_idle(layer_slicer, single_threaded_executor):
```

Comment:
This test is unrelated to changes in this PR but missed from previous.

Reply:
yes, this was a particularly tricky thing to test and we didn't come up with a good solution immediately.